### DISCUSSION TITLE
Accessing the last data in a buffer using SyncLogger class

### DISCUSSION THREAD

**Discussion by matteoviel18** [bitcraze/discussions#730] [General]
Hi everyone,

I'm currently working with the SyncLogger class and I'm trying to access the last available state of a drone. My goal is to get infos of the drone in real-time.

I've tried this solution (Python):

```
for uri in uris:
            last_data = None
            while True:
                if not state_loggers[uri]._queue.empty():  
                    last_data = state_loggers[uri].next()
                else:
                    break

            state_data_dict[uri] = last_data
```

where state_loggers[uri] is a SyncLogger object, state_data_dict[uri] is where I want to save the last state, .next() allows to empty the queue ( using _queue.get() ) and _queue.empty() should return True if the queue is empty, False otherwise.

However:
- in the documentation of _queue.empty(), the method is said to be not reliable
- I understand that my usage of _queue.empty() may not be the recommended approach, being _queue a protected member

So, I was wondering if a better solution should be used in this case.

Thank you in advance for your assistance!

Best regards,
Matteo

**Comment by krichardsson** [comment]
I think the best solution is to not use the SyncLogger at all. The SyncLogger was mainly intended to simplify scripts that sets up some log and perhaps print the values.
In your case a more appropriate solution would probably be to set up logging using a callback instead. You can find an example of how this is done in https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/logging/basiclog.py

Another way would be to use threads, but to me this feels convoluted. The purpose of the SyncLogger is to make an asynchronous process synchronous. By adding threads you would essentially make it asynchronous again, but then why not skip the SyncLogger all together?

**Reply by matteoviel18** [reply]
Thanks for the kind reply :)
